data-service/src/main/java/com/amazonaws/blox/dataservice/repository/model/EnvironmentDDBRecord.java

```diff
@@ -0,0 +1,96 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://aws.amazon.com/apache2.0/
+ *
+ * or in the "LICENSE" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazonaws.blox.dataservice.repository.model;
+
+import com.amazonaws.blox.dataservice.model.EnvironmentHealth;
+import com.amazonaws.blox.dataservice.model.EnvironmentStatus;
+import com.amazonaws.blox.dataservice.model.EnvironmentType;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBHashKey;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBIndexHashKey;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBIndexRangeKey;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBRangeKey;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBTable;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBTypeConverted;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBTypeConvertedEnum;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBVersionAttribute;
+import java.time.Instant;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@DynamoDBTable(tableName = "Environments")
+@Data
+@Builder
+//Required for builder because an empty constructor exists
+@AllArgsConstructor
+//Required for dynamodbmapper
+@NoArgsConstructor
+public class EnvironmentDDBRecord {
+
+  public static final String ENVIRONMENT_ID_HASH_KEY = "environmentId";
+  public static final String ENVIRONMENT_NAME_RANGE_KEY = "environmentName";
+  public static final String ENVIRONMENT_CLUSTER_GSI_NAME = "environmentClusterIndex";
+  public static final String ENVIRONMENT_CLUSTER_INDEX_HASH_KEY = "accountId";
+  public static final String ENVIRONMENT_CLUSTER_INDEX_RANGE_KEY = "cluster";
+
+  public static EnvironmentDDBRecord withKeys(
+      final String environmentId, final String environmentName) {
+
+    return EnvironmentDDBRecord.builder()
+        .environmentId(environmentId)
+        .environmentName(environmentName)
+        .build();
+  }
+
+  public static EnvironmentDDBRecord withGSIKeys(final String accountId, final String cluster) {
+    return EnvironmentDDBRecord.builder().accountId(accountId).clusterName(cluster).build();
+  }
+
+  @DynamoDBHashKey(attributeName = ENVIRONMENT_ID_HASH_KEY)
+  private String environmentId;
+
+  @DynamoDBRangeKey(attributeName = ENVIRONMENT_NAME_RANGE_KEY)
+  private String environmentName;
+
+  @DynamoDBIndexHashKey(
+    attributeName = ENVIRONMENT_CLUSTER_INDEX_HASH_KEY,
+    globalSecondaryIndexName = ENVIRONMENT_CLUSTER_GSI_NAME
```

Comment:
If I understand correctly, since we're re-using this POJO for the index, we'll get `EnvironmentDDBRecord` objects back from a query against the index, but with `null` values for all fields not projected into the index.

Perhaps it would make more sense to create a separate Record class (`EnvironmentClusterIndexDDBRecord`?) for the type of objects stored in the index? This is what [the docs suggest](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBMapper.Methods.html).

Reply:
Where does the doc suggest creating a separate object? IMO it creates additional complexity with having to map to a different object when creating/reading this record.

Reply:
Wouldn't it be confusing that a query for `listAllClusters` against the DB would return a set of `EnvironmentDDBRecord` objects?

For the docs, sorry, I missed the direct link to the section, [here](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBMapper.Methods.html#DynamoDBMapper.Methods.query) it is:

> To query an index, you must first model the index as a mapper class. Suppose that the Reply table has a global secondary index named `PostedBy-Message-Index`. The partition key for this index is `PostedBy`, and the sort key is `Message`. The class definition for an item in the index would look like this:

```java
@DynamoDBTable(tableName="Reply")
public class PostedByMessage { 
    private String postedBy;
    private String message;

    @DynamoDBIndexHashKey(globalSecondaryIndexName = "PostedBy-Message-Index", attributeName = "PostedBy")
    public String getPostedBy() { return postedBy; }
    public void setPostedBy(String postedBy) { this.postedBy = postedBy; } 

    @DynamoDBIndexRangeKey(globalSecondaryIndexName = "PostedBy-Message-Index", attributeName = "Message")
    public String getMessage() { return message; }
    public void setMessage(String message) { this.message = message; } 

   // Additional properties go here. 
}
```
> The `@DynamoDBTable` annotation indicates that this index is associated with the Reply table. The `@DynamoDBIndexHashKey` annotation denotes the partition key (PostedBy) of the index, and `@DynamoDBIndexRangeKey` denotes the sort key (Message) of the index.